3rdparty/ymfm/src/ymfm.h

```diff
@@ -0,0 +1,379 @@
+
+	// internal state
+	std::vector<uint8_t> &m_buffer;
+	int32_t m_offset;
+};
+
+
+
+//*********************************************************
+//  INTERFACE CLASSES
+//*********************************************************
+
+// ======================> ymfm_engine_callbacks
+
+// this class represents functions in the engine that the ymfm_interface
+// needs to be able to call; it is represented here as a separate interface
+// that is independent of the actual engine implementation
+class ymfm_engine_callbacks
+{
+public:
+	// timer callback; called by the interface when a timer fires
+	virtual void engine_timer_expired(uint32_t tnum) = 0;
+
+	// check interrupts; called by the interface after synchronization
+	virtual void engine_check_interrupts() = 0;
+
+	// mode register write; called by the interface after synchronization
+	virtual void engine_mode_write(uint8_t data) = 0;
+};
+
+
+// ======================> ymfm_interface
+
+// this class represents the interface between the fm_engine and the outside
+// world; it provides hooks for timers, synchronization, and IRQ signaling
+class ymfm_interface
+{
+	// the engine is our friend
+	template<typename RegisterType> friend class fm_engine_base;
+
+public:
+	// constructor
+	ymfm_interface() { }
+
+	// destructor
+	virtual ~ymfm_interface() { }
+
+	// logging helper
+	template<typename... Params>
+	void log(Params &&... args)
+	{
+		char buffer[256];
+		snprintf(buffer, sizeof(buffer), std::forward<Params>(args)...);
+		buffer[sizeof(buffer) - 1] = 0;
+		ymfm_log(buffer);
+	}
+
+	// the following functions must be implemented by any derived classes; the
+	// default implementations are sufficient for some minimal operation, but will
+	// likely need to be overridden to integrate with the outside world; they are
+	// all prefixed with ymfm_ to reduce the likelihood of namespace collisions
+
+	//
+	// timing and synchronizaton
+	//
+
+	// the chip implementation calls this when a write happens to the mode
+	// register, which could affect timers and interrupts; our responsibility
+	// is to ensure the system is up to date before calling the engine's
+	// engine_mode_write() method
+	virtual void ymfm_sync_mode_write(uint8_t data) { m_engine->engine_mode_write(data); }
+
+	// the chip implementation calls this when the chip's status has changed,
+	// which may affect the interrupt state; our responsibility is to ensure
+	// the system is up to date before calling the engine's
+	// engine_check_interrupts() method
+	virtual void ymfm_sync_check_interrupts() { m_engine->engine_check_interrupts(); }
+
+	// the chip implementation calls this when one of the two internal timers
+	// has changed state; our responsibility is to arrange to call the engine's
+	// engine_timer_expired() method after the provided number of clocks; if
+	// duration_in_clocks is negative, we should cancel any outstanding timers
+	virtual void ymfm_set_timer(uint32_t tnum, int32_t duration_in_clocks) { }
```

Comment:
Shouldn’t this be pure virtual if it actually needs a real implementation?

Reply:
There are cases (like a VGM player) where you don't need to deal with timers at all, so I left it with a default empty implementation rather than making it required. An emulator consumer would likely need to implement it of course.